Type models

vl0w opened the pull request:
# Summary
Support for models of arbitrary (local, non-local) types. A model represents an abstraction over a type that can be used in specifications.

# Motivation
Some structs, mostly non-local structs from other crates, contain non-public code which are essential for verification.
An example is the `std::slice::Iter` struct from the standard library. We would like to provide a specification for the `Iterator` implementation of `Iter`:
```rust
impl Iterator<'a, T> for std::slice::Iter<'a, T>{
    // spec???
    fn next(&mut self) -> Option<&'a T>;
}
```
There is not really a way to provide a concise specification for this iterator implementation, since `Iter`'s fields [do not allow a straightforward](https://doc.rust-lang.org/src/core/slice/iter.rs.html#65-71) specification.

We can instead provide a *model* `Iter` in the following way: The contained data and the current position in the iteration:
```rust
#[model]
struct std::slice::Iter<'a, T> where T: 'a {
    position: usize,
    data: &'a [T],
}
```
This model then can be used in the specification (just an outline, actual specs would look different):
```rust
impl Iterator<'a, T> for std::slice::Iter<'a, T>{
    #[ensures( self.model().data.len == old(self.model().data.len)) )]
    #[ensures( old(self.model().pos) < self.model().data.len ==> (result.is_some() && result.unwrap() == self.model().data[old(self.model().pos)]))]
    // ... more ...
    fn next(&mut self) -> Option<&'a T>;
}
```

# Implementation
This PR adds the following things:
- A procedural macro `#[model]` -> Please refer to the [code docs](https://github.com/vl0w/prusti-dev/blob/type-models/prusti-specs/src/type_model/mod.rs) and the [tests](https://github.com/vl0w/prusti-dev/blob/type-models/prusti-specs/src/type_model/mod.rs#L199) to see how it works
- An extension to `SpecChecker` to ensure that the `.model()` method gets only called in `#[prusti::spec_only]` annotated code.

vl0w committed:
Support for type models (initial)

vl0w committed:
Generate ToModel trait on the fly

vl0w committed:
Add comments, tests, initial support for generics

vl0w committed:
Adjust docs, add tests

vl0w committed:
Check that models are only used in specs

vl0w committed:
Adjust doc

vl0w committed:
Fix clippy errors

vl0w committed:
Format code

vl0w committed:
Format code

vl0w committed:
Update docs

Aurel300 commented on prusti-contracts-internal/src/lib.rs:
```diff
@@ -52,3 +52,40 @@ pub fn extern_spec(attr: TokenStream, tokens: TokenStream) -> TokenStream {
+/// * Specifications involving models can only be used in trusted functions or methods, i.e. on
+///   explicitly `#[trusted]` methods or on `#[extern_spec]`s.
+/// * Models can not be generic, i.e. creating a model for `Vec<T>` is not possible. It is however
+///   possible to model concrete generic types, i.e. `Vec<i32>` and `Vec<u32>`
```
Good documentation, but this would probably fit better in the user guide. I'm not sure docs show up correctly on Prusti attributes in the IDE. If yes, then it should be a short explanation + a link to the user guide.

vl0w replied:
I can at confirm that they don't show up in my IDE (clion). 
Maybe this becomes more relevant when `prusti_contracts` will be distributed crate on crates.io (#529)
For now, I will simply add this documentation to the user guide :+1:

vl0w replied:
PR #865 created

vl0w replied:
I incorporated your feedback in the original PR

Aurel300 commented on prusti-contracts/src/lib.rs:
```diff
@@ -78,8 +81,10 @@ mod private {
     pub use prusti_contracts_internal::predicate;
-}
 
+    /// Macro for creating type [models](prusti_contracts_internal::model).
```
```suggestion
    /// Macro for creating type models.
```
(Do re-export docs do anything in the IDE?)

Aurel300 commented on prusti-interface/src/specs/checker.rs:
```diff
@@ -126,12 +131,76 @@ impl<'v, 'tcx> Visitor<'tcx> for CheckPredicatesVisitor<'v, 'tcx> {
+    }
+
+    fn is_embedded_in_spec(&self, hir_id: hir::HirId) -> bool {
+        for (parent_local_def_id, _) in self.tcx.hir().parent_owner_iter(hir_id) {
```
Why is the approach here different from `CheckPredicatesVisitor`? That is, why do we iterate over parents for calls to model methods rather than simply not recursing into spec functions in the first place?

vl0w replied:
Good idea, I guess this will simplify the code

vl0w replied:
I did a somewhat bigger refactoring of `SpecChecker`. The checker for type models now behaves like `CheckPredicatesVisitor`.

Aurel300 commented on prusti-interface/src/utils.rs:
```diff
@@ -307,6 +307,10 @@ pub fn has_extern_spec_attr(attrs: &[ast::Attribute]) -> bool {
 }
 
+pub fn has_model_attr(attrs: &[ast::Attribute]) -> bool {
+    has_prusti_attr(attrs, "model_generator")
```
I don't like this name, we might be using Rust generators to model some things soonish :)

vl0w replied:
This is a simple marked used to check that `.model()` is called from spec code only. Suggestion for new name: `to_model_fn`?

Aurel300 replied:
`to_model_fn` or even just `to_model` should be good.

Aurel300 commented on prusti-specs/src/type_model/mod.rs:
```diff
@@ -0,0 +1,392 @@
+
+        GeneratedIdents {
+            model_struct_ident: Ident::new(
+                format!("Prusti{}Model", name).as_str(),
```
These should probably get some UUIDs to avoid conflicts.

Aurel300 commented on prusti-specs/src/type_model/mod.rs:
```diff
@@ -0,0 +1,392 @@
+        model_struct,
+        model_impl,
+        to_model_trait,
+    })
```
```suggestion
    Ok(TypeModel {
        model_struct: create_model_struct(&item_struct, &idents)?,
        model_impl: create_model_impl(&item_struct, &model_struct, &to_model_trait)?,
        to_model_trait: create_to_model_trait(&item_struct, &model_struct, &idents),
    })
```

Aurel300 commented on prusti-specs/src/type_model/mod.rs:
```diff
@@ -0,0 +1,392 @@
+
+    let model_struct_ident = &idents.model_struct_ident;
+    let mut model_struct: syn::ItemStruct = parse_quote_spanned! {item_struct.span()=>
+        #[derive(Copy, Clone)]
```
`PartialEq` and `Eq` as well?

vl0w replied:
Simply deriving those two traits will have no immediate effect since we cannot assume a specification for `PartialEq::eq`.
We could generate the specification during the macro expansion, but it might be overkill?

Aurel300 replied:
IIRC we treat types with derived `PartialEq` and `Eq` specially, because we use [`has_structural_eq_impls`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.has_structural_eq_impls)?

Aurel300 commented on prusti-specs/src/type_model/mod.rs:
```diff
@@ -0,0 +1,392 @@
+    fn from(err: TypeModelGenerationError) -> Self {
+        match err {
+            TypeModelGenerationError::MissingStructFields(span) => {
+                syn::Error::new(span, "Missing fields in model")
```
This error message is not great: it could be misunderstood as saying there are some *particular* fields which should be in the model. How about "Type model must have at least one field"?

vl0w replied:
Your suggestion is good :+1: 

If I recall correctly, there is no technical reason for this error. Imo, it just makes no sense to have a model without fields.

Aurel300 commented on prusti-tests/tests/typecheck/fail/type-model/generic-param.rs:
```diff
@@ -0,0 +1,12 @@
+use std::vec::Vec;
+
+#[model]
+struct Vec<T> { //~ ERROR: cannot find type `T` in this scope [E0412]
```
So the generics here are the instantiations? Is it not possible to define a type model for a generic type?

vl0w replied:
Currently it is not possible to define a model for a generic type.
It is, however, possible to define models for different concrete types (see [this test](https://github.com/vl0w/prusti-dev/blob/650ac711ca2e1af82144cdf560086e00a27954b1/prusti-tests/tests/verify/pass/type-model/models-for-different-monomorphizations.rs)).

To support generic type parameters as well, we would need annotations whether we refer to a _concrete_ or _generic_ type as we did in extern trait specs (e.g. [here](https://github.com/vl0w/prusti-dev/blob/extern-trait-specs/prusti-tests/tests/verify/pass/extern-spec/traits/generics-1.rs) or [here](https://github.com/vl0w/prusti-dev/blob/extern-trait-specs/prusti-tests/tests/verify/pass/extern-spec/traits/generics-2.rs))

vl0w replied:
It might be necessary to add this feature, because we want to have a generic model for `std::iter::Iter`

vl0w replied:
This feature should be implemented by [e76b3ac](https://github.com/viperproject/prusti-dev/pull/836/commits/e76b3ac6f4a0791f2601cb0a1fe340a7f78eda7e)

vl0w committed:
Adjust ModelUsageVisitor to comply with latest rustc version

vl0w committed:
Update docs

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Small refactoring of type_model/mod.rs

Update prusti-specs/src/type_model/mod.rs (

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

vl0w committed:
Refactoring of type model macro
- Use UUID for generated struct and trait
- Rename 'model_generator' annotation

vl0w committed:
Remove macro documentation (moved to user guide)

vl0w committed:
Add user guide for type models

vl0w committed:
Refactoring of SpecChecker

vl0w committed:
Ignore warning due to randomized name

vl0w committed:
Check that modelled types have fields

Emit a warning otherwise

vl0w committed:
Fix wrong attribute

vl0w committed:
Rename directory

vl0w committed:
Adjust docs

Aurel300 commented on prusti-tests/tests/verify/pass/type-model/models-for-same-type-unsound.rs:
```diff
@@ -0,0 +1,35 @@
+// Hence, we cannot refer to different models for different instances of `A`.
+// If `A` had fields, the code verifies as expected.
+//
+// This code will print a warning to a user.
```
If this test is meant to check that the warning is printed, we can put it into `verify(_overflow)/ui`.

vl0w replied:
There is a separate test which checks the warning in `verify/ui`.
This test is meant as a "documentation" of the behavior. It is maybe not needed, but I found it somewhat useful. Mentioning it in the user docs might be good enough though.

Aurel300 replied:
Ok, then I would delete this and only keep the `ui` test. As it is, it is just testing the same thing again, without checking the warning. Oh also, by default tests should be in `verify_overflow`, as that is the default in Prusti.

vl0w replied:
Test deleted, other tests moved

vl0w committed:
Move type model tests

vl0w committed:
Add generic support for type models

vl0w committed:
Adjust type model user docs

vl0w committed:
Post rebase fixes
